### PR TITLE
Allow literal types without `Literal`

### PR DESCRIPTION
**Pull request by KotlinIsland**
`Literal` is literally an impostor.

- partially resolves:  #1

### CODE REVIEW

File: mypy/typeanal.py [hunk 1]
```diff
@@ -784,7 +785,13 @@ def visit_raw_expression_type(self, t: RawExpressionType) -> Type:
         # make signatures like "foo(x: 20) -> None" legal, we can change
         # this method so it generates and returns an actual LiteralType
         # instead.
```
Comment (DetachHead) [inline mypy/typeanal.py:787]: this comment is wrong now

File: mypy/typeanal.py [hunk 2]
```diff
@@ -784,7 +785,13 @@ def visit_raw_expression_type(self, t: RawExpressionType) -> Type:
...
-
+        # Sounds based to me
+        if (
+            self.api.is_future_flag_set("annotations")
+            and t.base_type_name in ('builtins.int', 'builtins.bool')
```
Comment (DetachHead) [inline mypy/typeanal.py:791]: support `str` and just ban quoted types when `__future__.annotations` is set

File: test-data/unit/check-based-literal-types.test [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+[builtins fixtures/tuple.pyi]
+
+[case testNoBareLiteralEnum]
+from enum import Enum
+
+class E(Enum):
+    A = 1
+
+a: E.A  # E: Invalid type: try using Literal[E.A] instead?  [misc]
+[builtins fixtures/tuple.pyi]
```
Comment (DetachHead) [inline test-data/unit/check-based-literal-types.test:51]: add a test for unions of literals? eg. `True | 2`

File: mypy/typeanal.py (hunk 2 above)
Comment (KotlinIsland) [reply mypy/typeanal.py:791]: that sounds like more of a disruptive change, maybe in a follow-up

Comment (DetachHead) [reply mypy/typeanal.py:788]: #361

**Review by DetachHead — APPROVED** [review]

File: mypy/types.py [hunk 4]
```diff
@@ -756,6 +757,9 @@ def __init__(self,
         self.original_str_expr = original_str_expr
         self.original_str_fallback = original_str_fallback
 
+        # this is used to ban bare Enum literals from TypeAliases
+        self.from_alias = from_alias
```
Comment (DetachHead) [inline mypy/types.py:761]:
```suggestion
        self.from_alias = from_alias
        """this is used to ban bare Enum literals from TypeAliases"""
```

File: test-data/unit/check-based-literal-types.test [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+[case testNoBareLiteralEnum]
+from enum import Enum
+
+class E(Enum):
+    A = 1
+
+a: E.A
+a: E.A | str  # E: Invalid type: try using Literal[E.A] instead?  [misc]
```
Comment (DetachHead) [inline test-data/unit/check-based-literal-types.test:61]: shouldn't this error be `valid-type`

File: test-data/unit/check-based-literal-types.test [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+class E(Enum):
+    A = 1
+
+A: TypeAlias = 1  # E: Invalid type alias: expression is not a valid type  [misc]
+B: TypeAlias = True  # E: Invalid type alias: expression is not a valid type  [misc]
+C: TypeAlias = False  # E: Invalid type alias: expression is not a valid type  [misc]
+D: TypeAlias = False  # E: Invalid type alias: expression is not a valid type  [misc]
+E: TypeAlias = E.A  # E: Invalid type alias: expression is not a valid type  [misc]
+[builtins fixtures/tuple.pyi]
```
Comment (DetachHead) [inline test-data/unit/check-based-literal-types.test:79]: add test(s) for quoted types?

File: test-data/unit/check-based-literal-types.test (hunk 5 above)
Comment (KotlinIsland) [reply test-data/unit/check-based-literal-types.test:61]: uhhh, yeah lol

Comment (KotlinIsland) [reply test-data/unit/check-based-literal-types.test:61]: but that's a breaking change, I don't want to touch it here, maybe up stream.

Comment (DetachHead) [reply test-data/unit/check-based-literal-types.test:61]: isn't that the point of running tests in based mode and non-based mode?

Comment (KotlinIsland) [reply test-data/unit/check-based-literal-types.test]: nah, it's like an existing error that is broken upstream

File: test-data/unit/check-based-literal-types.test (hunk 6 above)
Comment (KotlinIsland) [reply test-data/unit/check-based-literal-types.test:82]: to be fair quoted types in aliases are already extremely busdted, so I don't think I'll go near it for now.

**Comment by KotlinIsland** [comment]
@DetachHead These changes https://github.com/KotlinIsland/basedmypy/compare/73b1142..081a0778e4a72bf63439856a680ded7a81de0a12

File: mypy/typeanal.py [hunk 7]
```diff
@@ -800,14 +800,31 @@ def visit_raw_expression_type(self, t: RawExpressionType) -> Type:
...
-                self.note(t.note, t, code=codes.VALID_TYPE)
-
+            if msg:
+                self.fail(msg, t, code=codes.VALID_TYPE)
+                if t.note is not None:
+                    self.note(t.note, t, code=codes.VALID_TYPE)
+        if t.base_type_name in ('builtins.int', 'builtins.bool'):
+            v = t.literal_value
+            assert v is not None
+            return LiteralType(
```
[the remark this answers is not in the document]
Comment (KotlinIsland) [reply mypy/typeanal.py:810]: nvm

**Pull request merged**